### PR TITLE
[CF-558] Add API call retrying to discover process

### PR DESCRIPTION
**Pull request by antonf**
Before this commit any random HTTP error (like 503) could ruin
one hour discover run. This commit add API call retrying, which
mean that in case of some non-fatal random error API request will
be repeated configured amount of times after configured amount
of time.

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3869/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.discovery.nova
cloudferry/lib/os/discovery/nova.py:49: [E1101(no-member), 
FlavorDiscoverer.discover_one] Instance of 'list' has no 'id' member

**Comment by MirantisWorkloadMobility** [comment]
$PYLINT_STATUS

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3545/

**Comment by MirantisWorkloadMobility** [comment]
Merge check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3870/

**Comment by MirantisWorkloadMobility** [comment]
$PYLINT_STATUS

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.discover
cloudferry/discover.py:15: [W0611(unused-import), ] Unused import collections
************* Module cloudferry.lib.os.discovery.nova
cloudferry/lib/os/discovery/nova.py:49: [E1101(no-member), 
FlavorDiscoverer.discover_one] Instance of 'list' has no 'id' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3546/

**Comment by MirantisWorkloadMobility** [comment]
$PYLINT_STATUS

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.discovery.nova
cloudferry/lib/os/discovery/nova.py:49: [E1101(no-member), 
FlavorDiscoverer.discover_one] Instance of 'list' has no 'id' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
$PYLINT_STATUS

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by antonf** [comment]
retest

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: punky_gibbon/install_punky_gibbon.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+LOCAL_IP=${1}
+RANDOM_CHANCE=${2}
+API_PASTE_CONFIGS="/etc/nova/api-paste.ini /etc/glance/glance-api-paste.ini /etc/cinder/api-paste.ini"
+SERVICES="nova-api glance-api cinder-api"
+
+install_punky_gibbon() {
+    FILE=${1}
+    SECTION=${2}
+    OPTION=${3}
```
Comment (roman-verchikov) [inline punky_gibbon/install_punky_gibbon.sh]: please use `local` variables inside functions

File: punky_gibbon/install_punky_gibbon.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        elif [[ ${section} == composite:* && $(crudini --get "${config_file}" "${section}" use) == *:pipeline_factory ]]; then
+            for option in $(crudini --get "${config_file}" "${section}" | grep -v use); do
+                install_punky_gibbon "${config_file}" "${section}" "${option}"
+            done
+        fi
+    done
+
+    crudini --set "${FILE}" "filter:punky_gibbon" "paste.filter_app_factory" "punky_gibbon.middleware:PunkyGibbon"
+    crudini --set "${FILE}" "filter:punky_gibbon" "ignore_ips" "127.0.0.1,${LOCAL_IP}"
+    crudini --set "${FILE}" "filter:punky_gibbon" "random_chance" "${RANDOM_CHANCE}"
```
Comment (roman-verchikov) [inline punky_gibbon/install_punky_gibbon.sh]: would be safer to use `${config_file}` here

File: punky_gibbon/punky_gibbon/middleware.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+class PunkyGibbon(object):
+    def __init__(self, app, global_conf, random_chance=0.5, ignore_ips='',
+                 message=None, detailed_message=None, **kwargs):
+        self.app = app
+        self.global_conf = global_conf
+        self.random_chance = float(random_chance)
+        self.message = message or DEFAULT_MESSAGE
+        self.detailed_message = detailed_message or DEFAULT_MESSAGE
+        self.ignore_ips = [ip.strip() for ip in ignore_ips.split(',')]
+        LOG.error('Ignored ips: %s %s', self.ignore_ips, kwargs)
```
Comment (roman-verchikov) [inline punky_gibbon/punky_gibbon/middleware.py]: I guess `INFO` level should be enough here

**Comment by roman-verchikov** [comment]
Looks good other than several minor comments, please resolve

**Comment by MirantisWorkloadMobility** [comment]
$PULL_STATUS
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3874/

**Comment by MirantisWorkloadMobility** [comment]
$PYLINT_STATUS

**Comment by roman-verchikov** [comment]
retest

**Comment by roman-verchikov** [comment]
+1

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module punky_gibbon.middleware
punky_gibbon/punky_gibbon/middleware.py:23: [W0613(unused-argument), 
PunkyGibbon.__init__] Unused argument 'kwargs'

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by antonf** [comment]
retest

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/config.py [hunk 4]
```diff
@@ -289,6 +289,8 @@ class DatabaseSettings(ConfigSection):
 
 class OpenstackCloud(ConfigSection):
     name = fields.String()
+    request_attempts = fields.Integer(missing=3)
```
Comment (archyufa) [inline cloudferry/config.py]: Can you add comment what is the parameter does?

File: cloudferry/config.py [hunk 5]
```diff
@@ -289,6 +289,8 @@ class DatabaseSettings(ConfigSection):
...
+    request_failure_sleep = fields.Float(missing=5)
```
Comment (archyufa) [inline cloudferry/config.py]: Can you add comment what is the parameter does?

**Comment by archyufa** [comment]
+1 tested in staging

**Comment by antonf** [comment]
retest

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/config.py (hunk 4 above)
Comment (antonf) [inline cloudferry/config.py]: No, I can't. I still need to figure out how to generate configuration documentation from config, so It is separate task.

File: cloudferry/config.py (hunk 5 above)
Comment (antonf) [inline cloudferry/config.py]: No, I can't. I still need to figure out how to generate configuration documentation from config, so It is separate task.

**Comment by roman-verchikov** [comment]
Can be merged

**Pull request merged**